packages/core/src/scp-cf/destination-accessor.ts

```diff
@@ -482,19 +285,346 @@ const emptyDestinationByType: DestinationsByType = {
   subaccount: []
 };
 
-function subscriberDestinationIsSelected(
-  selectionStrategy: DestinationSelectionStrategy
-): boolean {
-  return (
-    selectionStrategy === subscriberFirst ||
-    selectionStrategy === alwaysSubscriber
-  );
-}
-
 function emptyDestinationsByType(
   destinationByType: DestinationsByType
 ): boolean {
   return (
-    !destinationByType.instance.length && !destinationByType.instance.length
+    !destinationByType.instance.length && !destinationByType.subaccount.length
   );
 }
+
+class DestinationAccessor {
+  public static async getDestinationFromDestinationService(
+    name: string,
+    options: DestinationOptions & { iss?: string }
+  ): Promise<Destination | undefined> {
+    const decodedUserJwt = await DestinationAccessor.getDecodedUserJwt(options);
+    const providerToken = await DestinationAccessor.getProviderToken(options);
+    const da = new DestinationAccessor(
+      name,
+      options,
+      decodedUserJwt,
+      providerToken
+    );
+
+    return (
+      da.trySubscriberCache() ??
+      (await da.tryProviderCache()) ??
+      (await da.processDestinationResult(await da.tryForOAuth())) ??
+      (await da.processDestinationResult(await da.tryForCert())) ??
+      (await da.processDestinationResult(await da.tryService())) ??
+      undefined
+    );
+  }
+
+  private static async getDecodedUserJwt(
+    options: DestinationOptions & { iss?: string }
+  ): Promise<DecodedJWT | undefined> {
+    return options.userJwt
+      ? verifyJwt(options.userJwt, options)
+      : options.iss
+      ? { iss: options.iss }
+      : undefined;
+  }
+
+  private static async getProviderToken(
+    options: DestinationOptions
+  ): Promise<string> {
+    const { userJwt, ...optionsWithoutJwt } = options;
+    return serviceToken('destination', optionsWithoutJwt);
+  }
+
+  private subscriberDestinationCache: DestinationsByType | undefined;
+  private providerDestinationCache: DestinationsByType | undefined;
+
+  private constructor(
+    private name: string,
+    private options: DestinationOptions,
+    private decodedUserJwt: DecodedJWT | undefined,
+    private providerToken: string
```

Comment:
strongly consider making these `readonly`

Reply:
I would love: private and readonly because I do not want to expose them and also make them not changeable and be forced to initialize them. But I guess readonly is the more modern and more used keyword.